File: taint-fixtures-v2/java/cwe22/TpZipSlipNoValidation.java
// Fixture: code_patch · CWE-22 Path Traversal · Java
// VERDICT: TRUE_POSITIVE
// PATTERN: zip_extract_no_path_check
// SOURCE: zip_entry_name
// SINK: Files.copy
// TAINT_HOPS: 1
// NOTES: Classic zip-slip — no path validation after resolve
import java.io.*;
import java.nio.file.*;
import java.util.zip.*;
import java.util.Enumeration;

public class TpZipSlipNoValidation {
    public void extractUnsafe(File zipFile, File outputDir) throws Exception {
        try (ZipFile zip = new ZipFile(zipFile)) {
            Enumeration<? extends ZipEntry> entries = zip.entries();
            while (entries.hasMoreElements()) {
                ZipEntry entry = entries.nextElement();
                // VULNERABLE: no startsWith check — ../../etc/passwd escapes
                File outputFile = new File(outputDir, entry.getName());
                Files.copy(zip.getInputStream(entry), outputFile.toPath());
            }
        }
    }
}

File: taint-fixtures-v2/go/cwe22/tp_zip_slip_no_validation.go
// Fixture: code_patch · CWE-22 Path Traversal · Go
// VERDICT: TRUE_POSITIVE
// PATTERN: zip_slip_no_path_validation
// SOURCE: archive (zip entry name)
// SINK: os.Create (unvalidated path)
// TAINT_HOPS: 1
// NOTES: Classic zip-slip — Fleet TP pattern (update.go:754)
package extractor

import (
	"archive/zip"
	"io"
	"os"
	"path/filepath"
)

func ExtractZipUnsafe(zipPath, destDir string) error {
	r, err := zip.OpenReader(zipPath)
	if err != nil {
		return err
	}
	defer r.Close()

	for _, f := range r.File {
		// VULNERABLE: CWE-22 · no validation of zip entry name
		fpath := filepath.Join(destDir, f.Name)
		rc, _ := f.Open()
		outFile, _ := os.Create(fpath)
		io.Copy(outFile, rc)
		outFile.Close()
		rc.Close()
	}
	return nil
}
